flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -525,63 +525,88 @@ class ExprLowering {
-          return genRealConstant<KIND>(builder.getContext(), floatVal);
-        } else if constexpr (KIND == 10) {
-          llvm::APFloat floatVal{llvm::APFloatBase::x87DoubleExtended(), str};
-          return genRealConstant<KIND>(builder.getContext(), floatVal);
-        } else if constexpr (KIND == 16) {
-          llvm::APFloat floatVal{llvm::APFloatBase::IEEEquad(), str};
-          return genRealConstant<KIND>(builder.getContext(), floatVal);
-        } else {
-          // convert everything else to double
-          llvm::APFloat floatVal{llvm::APFloatBase::IEEEdouble(), str};
-          return genRealConstant<KIND>(builder.getContext(), floatVal);
-        }
+      std::string str = value.DumpHexadecimal();
+      if constexpr (KIND == 2) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEhalf(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 4) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEsingle(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 10) {
+        llvm::APFloat floatVal{llvm::APFloatBase::x87DoubleExtended(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 16) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEquad(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else {
+        // convert everything else to double
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEdouble(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
       }
-      llvm_unreachable("real constant has no value");
     } else if constexpr (TC == Fortran::lower::ComplexCat) {
-      auto opt = con.GetScalarValue();
-      if (opt.has_value()) {
-        using TR = Fortran::evaluate::Type<Fortran::lower::RealCat, KIND>;
-        return genval(Fortran::evaluate::ComplexConstructor<KIND>{
-            Fortran::evaluate::Expr<TR>{
-                Fortran::evaluate::Constant<TR>{opt->REAL()}},
-            Fortran::evaluate::Expr<TR>{
-                Fortran::evaluate::Constant<TR>{opt->AIMAG()}}});
-      }
-      llvm_unreachable("array of complex unhandled");
-    } else if constexpr (TC == Fortran::lower::CharacterCat) {
-      return genCharLit<KIND>(con.GetScalarValue().value(), con.LEN());
+      using TR = Fortran::evaluate::Type<Fortran::lower::RealCat, KIND>;
+      return genval(Fortran::evaluate::ComplexConstructor<KIND>{
+          Fortran::evaluate::Expr<TR>{
+              Fortran::evaluate::Constant<TR>{value.REAL()}},
+          Fortran::evaluate::Expr<TR>{
+              Fortran::evaluate::Constant<TR>{value.AIMAG()}}});
     } else {
       llvm_unreachable("unhandled constant");
     }
   }
 
+  template <Fortran::common::TypeCategory TC, int KIND>
+  mlir::Value genArrayLit(
```

Comment:
Is "Lit" a convention used elsewhere in llvm and mlir?   I'm not totally against it.

Reply:
I just followed the "Lit" naming used by "genCharLit()" which was instantiating a global char constant.
I'm not aware any "Lit" convention used in llvm or mlir.

Reply:
I see occasional uses of lit (for compile time constants) in MLIR in different forms.

```
auto *structLit = llvm::cast<StructLiteralExprAST>(var.get());
```

```
} else if (auto *lit = llvm::dyn_cast<LiteralExprAST>(var.get())) {
```
But I wouldn't call it a naming convention that they follow. All it says is `Lit` seems okay, since it shows up in MLIR too.